storm-core/src/jvm/org/apache/storm/zookeeper/Zookeeper.java

```diff
@@ -0,0 +1,355 @@
+        try {
+            if (watch) {
+                stat = zk.checkExists().watched().forPath(Utils.normalizePath(path));
+            } else {
+                stat = zk.checkExists().forPath(Utils.normalizePath(path));
+            }
+        } catch (Exception e) {
+            throw Utils.wrapInRuntime(e);
+        }
+
+        return stat != null;
+    }
+
+    public static void deleteNode(CuratorFramework zk, String path) throws RuntimeException {
+        try {
+            String npath = Utils.normalizePath(path);
+            if (existsNode(zk, npath, false)) {
+                zk.delete().deletingChildrenIfNeeded().forPath(Utils.normalizePath(path));
+            }
+
+        } catch (KeeperException.NoNodeException e) {
+            LOG.info("exception", e);
+        } catch (Exception e) {
+            throw Utils.wrapInRuntime(e);
+        }
+    }
+
+    public static void mkdirs(CuratorFramework zk, String path, List<ACL> acls) throws RuntimeException {
+
+        String npath = Utils.normalizePath(path);
+        if (npath.equals("/")) {
+            return;
+        }
+        if (existsNode(zk, npath, false)) {
+            return;
+        }
+        byte[] byteArray = new byte[1];
+        byteArray[0] = (byte) 7;
+        createNode(zk, npath, byteArray, org.apache.zookeeper.CreateMode.PERSISTENT, acls);
+
+    }
+
+    public static void syncPath(CuratorFramework zk, String path) throws RuntimeException {
+        try {
+            zk.sync().forPath(Utils.normalizePath(path));
+        } catch (Exception e) {
+            throw Utils.wrapInRuntime(e);
+        }
+    }
+
+    public static void addListener(CuratorFramework zk, ConnectionStateListener listener) {
+        zk.getConnectionStateListenable().addListener(listener);
+    }
+
+    public static byte[] getData(CuratorFramework zk, String path, boolean watch) throws RuntimeException {
+
+        try {
+            String npath = Utils.normalizePath(path);
+            if (existsNode(zk, npath, watch)) {
+                if (watch) {
+                    return zk.getData().watched().forPath(npath);
+                } else {
+                    return zk.getData().forPath(npath);
+                }
+            }
+        } catch (KeeperException e) {
+            // this is fine b/c we still have a watch from the successful exists call
+        } catch (Exception e) {
+            throw Utils.wrapInRuntime(e);
+        }
+
+        return null;
+    }
+
+    public static Integer getVersion(CuratorFramework zk, String path, boolean watch) throws Exception {
+        String npath = Utils.normalizePath(path);
+        Stat stat = null;
+        if (existsNode(zk, npath, watch)) {
+            if (watch) {
+                stat = zk.checkExists().watched().forPath(Utils.normalizePath(path));
+            } else {
+                stat = zk.checkExists().forPath(Utils.normalizePath(path));
+            }
+            return Integer.valueOf(stat.getVersion());
```

Comment:
There is potentially a race condition here that is not in the previous code.  I know this is minor but the checkExists() was there to eliminate this race.  I think what we want for this function is something more like.
```
public static Integer getVersion(CuratorFramework zk, String path, boolean watch) throws Exception {
    String npath = Utils.normalizePath(path);
    Stat stat = null;
    if (watch) {
        stat = zk.checkExists().watched().forPath(Utils.normalizePath(path));
    } else {
         stat = zk.checkExists().forPath(Utils.normalizePath(path));
     }
    return stat == null ? null : Integer.valueOf(stat.getVersion());
}
```